--- a/ipq40xx-adss.c
+++ b/ipq40xx-adss.c
@@ -39,7 +39,7 @@
 static spinlock_t glb_mode_lock;
 
 /* I2S Interface Enable */
-void ipq40xx_glb_i2s_interface_en(int enable)
+static void ipq40xx_glb_i2s_interface_en(int enable)
 {
 	uint32_t cfg;
 	unsigned long flags;
@@ -55,7 +55,7 @@
 }
 
 /* Enable Stereo0/Stereo1/Stereo2 channel */
-void ipq40xx_glb_stereo_ch_en(int enable, int stereo_ch)
+static void __maybe_unused ipq40xx_glb_stereo_ch_en(int enable, int stereo_ch)
 {
 	uint32_t cfg;
 	unsigned long flags;
@@ -77,7 +77,7 @@
 }
 
 /* I2S Module Reset */
-void ipq40xx_glb_i2s_reset(uint32_t reset)
+static void ipq40xx_glb_i2s_reset(uint32_t reset)
 {
 	writel(GLB_I2S_RESET_VAL, adss_audio_local_base + ADSS_GLB_I2S_RST_REG);
 	mdelay(5);
@@ -265,7 +265,7 @@
 	spin_unlock_irqrestore(&tdm_ctrl_lock, flags);
 }
 
-void ipq40xx_audio_adss_init(void)
+static void ipq40xx_audio_adss_init(void)
 {
 	spin_lock_init(&i2s_ctrl_lock);
 	spin_lock_init(&tdm_ctrl_lock);
